File: projects/11-water-treatment-utility/backend/dashboard/lambdaHandler.js
"use strict";

// AWS Lambda entry point for the dashboard behind an API Gateway REST API.
// Dispatch here is a single switch(true) statement whose cases are compound
// method-and-path boolean expressions -- a distinct shape from every other
// project's dashboard router (flat method+path map, template-segment matcher,
// ordered regex list, string/pattern switch, enum/sealed registry, chain of
// responsibility, or an Express/WSGI/Mangum bridge). It reuses the same data
// functions the local HTTP server (server.js) calls, so both front doors serve
// identical responses from one set of query logic, with a wildcard cross-origin
// header on every response so the S3-hosted frontend can call it cross-origin.

const { buildClients } = require("./awsClients");
const {
  SENSOR_TYPES,
  latestWindowsFor,
  buildPlantSummaries,
  getPlantSummary,
  freshestAgeSeconds,
} = require("./readingsStore");
const {
  isQueueReachable,
  isLambdaActive,
  readQueueCounters,
  countTableItems,
  checkGateway,
  isPipelineFlowing,
} = require("./pipelineStatus");
const { fetchThresholds } = require("./thresholdsProxy");

const TABLE_NAME = process.env.TABLE_NAME || "wtu-readings";
const QUEUE_NAME = process.env.SQS_QUEUE_NAME || "wtu-plant-agg";
const FUNCTION_NAME = process.env.LAMBDA_FUNCTION_NAME || "wtu-processor";
const FOG_HEALTH_URL = process.env.FOG_HEALTH_URL || "http://fog:8000/health";
const FOG_THRESHOLDS_URL = process.env.FOG_THRESHOLDS_URL || "http://fog:8000/thresholds";

const CORS = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "GET,OPTIONS",
  "Access-Control-Allow-Headers": "*",
};

// Clients and derived deps are built once per container and reused across
// invocations. A pre-built clients object may be injected for unit tests.
let cached = null;
function getDeps(injected) {
  if (injected) return depsFrom(injected);
  if (!cached) cached = depsFrom(buildClients());
  return cached;
}
function depsFrom(clients) {
  return {
    doc: clients.doc,
    sqs: clients.sqs,
    lambda: clients.lambda,
    tableName: TABLE_NAME,
    queueName: QUEUE_NAME,
    functionName: FUNCTION_NAME,
    gatewayHealthUrl: FOG_HEALTH_URL,
    gatewayThresholdsUrl: FOG_THRESHOLDS_URL,
  };
}

// Each endpoint returns a plain {status, body}; the caller renders it.
async function readings(d, query) {
  const sensorType = query.sensor_type;
  const limit = parseInt(query.limit || "60", 10);
  if (!sensorType || !SENSOR_TYPES.includes(sensorType)) {
    return { status: 400, body: { error: `sensor_type must be one of ${SENSOR_TYPES.join(", ")}` } };
  }
  const siteId = query.site_id;
  let items = await latestWindowsFor(d.doc, d.tableName, sensorType, siteId ? limit * 4 : limit);
  if (siteId) items = items.filter((item) => item.site_id === siteId).slice(-limit);
  return { status: 200, body: { sensor_type: sensorType, items } };
}

async function plants(d) {
  return { status: 200, body: { plants: await buildPlantSummaries(d.doc, d.tableName) } };
}

async function plantDetail(d, plantId) {
  const plant = await getPlantSummary(d.doc, d.tableName, plantId);
  if (!plant) return { status: 404, body: { error: `unknown plant: ${plantId}` } };
  return { status: 200, body: plant };
}

async function health(d) {
  const [gateway, queue, lambda, freshestAge] = await Promise.all([
    checkGateway(d.gatewayHealthUrl),
    isQueueReachable(d.sqs, d.queueName),
    isLambdaActive(d.lambda, d.functionName),
    freshestAgeSeconds(d.doc, d.tableName),
  ]);
  return { status: 200, body: { gateway, queue, lambda, pipeline: isPipelineFlowing(freshestAge), freshest_age_seconds: freshestAge } };
}

async function backendStats(d) {
  const [queue, itemsInTable] = await Promise.all([
    readQueueCounters(d.sqs, d.queueName),
    countTableItems(d.doc, d.tableName),
  ]);
  return { status: 200, body: { queue, items_in_table: itemsInTable } };
}

async function thresholds(d) {
  const result = await fetchThresholds(d.gatewayThresholdsUrl);
  return { status: result.status, body: result.body };
}

function respond(status, body) {
  return {
    statusCode: status,
    headers: { "Content-Type": "application/json", ...CORS },
    body: typeof body === "string" ? body : JSON.stringify(body),
  };
}

async function handler(event, _context, injectedClients) {
  const method = event.httpMethod || "GET";
  const path = event.path || "/";
  const query = event.queryStringParameters || {};
  if (method === "OPTIONS") return { statusCode: 200, headers: CORS, body: "" };

  const d = getDeps(injectedClients);
  const plantPath = /^\/api\/plants\/([a-z0-9-]+)$/.exec(path);
  try {
    let result;
    switch (true) {
      case method === "GET" && path === "/api/readings":
        result = await readings(d, query); break;
      case method === "GET" && path === "/api/plants":
        result = await plants(d); break;
      case method === "GET" && plantPath !== null:
        result = await plantDetail(d, plantPath[1]); break;
      case method === "GET" && path === "/api/health":
        result = await health(d); break;
      case method === "GET" && path === "/api/backend-stats":
        result = await backendStats(d); break;
      case method === "GET" && path === "/api/thresholds":
        result = await thresholds(d); break;
      default:
        result = { status: 404, body: { error: "not found" } };
    }
    return respond(result.status, result.body);
  } catch (err) {
    return respond(500, { error: err.message || "internal error" });
  }
}

module.exports = { handler };

File: projects/21-bridge-structural-health/backend/dashboard/lambda_handler.py
"""API Gateway (REST API, proxy integration) entry point for the dashboard.
Dispatches on (method, path) via a single PEP 634 structural-pattern match
statement -- Python's own match/case, not a dict lookup, a decorator
registry, a regex scan, or a class hierarchy. Reuses data_access.py and
thresholds_proxy.py directly rather than duplicating their logic."""

import json
import os
from datetime import datetime, timezone

import data_access
from thresholds_proxy import ThresholdsUnavailable, fetch_thresholds

FOG_HEALTH_URL = os.getenv("FOG_HEALTH_URL", "http://fog:8000/health")
FOG_THRESHOLDS_URL = os.getenv("FOG_THRESHOLDS_URL", "http://fog:8000/thresholds")
PIPELINE_FRESH_SECONDS = float(os.getenv("PIPELINE_FRESH_SECONDS", "30"))

CORS_HEADERS = {"Access-Control-Allow-Origin": "*", "Content-Type": "application/json"}


def _response(status, body):
    return {"statusCode": status, "headers": CORS_HEADERS, "body": json.dumps(body)}


def _fog_reachable():
    try:
        fetch_thresholds(FOG_HEALTH_URL)
        return True
    except ThresholdsUnavailable:
        return False


def _readings(params):
    sensor_type = params.get("sensor_type", "")
    if sensor_type not in data_access.SENSOR_TYPES:
        return _response(400, {"error": f"unknown sensor_type: {sensor_type}"})

    try:
        limit = int(params.get("limit", "60"))
        if limit <= 0:
            raise ValueError
    except ValueError:
        return _response(400, {"error": "limit must be a positive integer"})

    site_id = params.get("site_id")
    items = data_access.recent_windows(sensor_type, limit)
    if site_id:
        items = [item for item in items if item["site_id"] == site_id]
    return _response(200, {"sensor_type": sensor_type, "items": items})


def _spans(_params):
    return _response(200, {"spans": data_access.span_report()})


def _thresholds(_params):
    try:
        payload = fetch_thresholds(FOG_THRESHOLDS_URL)
    except ThresholdsUnavailable as exc:
        return _response(502, {"error": str(exc)})
    return _response(200, payload)


def _health(_params):
    freshest_age = data_access.freshest_window_age(datetime.now(timezone.utc))
    pipeline_ok = freshest_age is not None and freshest_age <= PIPELINE_FRESH_SECONDS
    return _response(200, {
        "gateway": _fog_reachable(),
        "queue": data_access.queue_reachable(),
        "lambda": data_access.lambda_active(),
        "pipeline": pipeline_ok,
        "freshest_age_seconds": freshest_age,
    })


def _backend_stats(_params):
    return _response(200, {
        "queue": data_access.queue_depth(),
        "items_in_table": data_access.items_in_table(),
    })


def lambda_handler(event, _context):
    method = event.get("httpMethod", "GET")
    path = event.get("path", "/")
    params = event.get("queryStringParameters") or {}

    if method == "OPTIONS":
        return _response(200, {})

    match (method, path):
        case ("GET", "/api/readings"):
            return _readings(params)
        case ("GET", "/api/spans"):
            return _spans(params)
        case ("GET", "/api/thresholds"):
            return _thresholds(params)
        case ("GET", "/api/health"):
            return _health(params)
        case ("GET", "/api/backend-stats"):
            return _backend_stats(params)
        case _:
            return _response(404, {"error": f"no route for {method} {path}"})
